File: frontend/src/services/api/focus.ts
import { apiClient } from './client';
import { Task } from '@/types';

export interface FocusSession {
  id: string;
  user_id: string;
  task_id?: string;
  started_at: string;
  ended_at?: string;
  duration_seconds: number;
  productive_seconds: number;
  distracted_seconds: number;
  status: 'RUNNING' | 'PAUSED' | 'COMPLETED' | 'CANCELLED';
  created_at: string;
  task?: Task;
}

export interface FocusSummaryToday {
  total_sessions: number;
  focus_seconds: number;
  distracted_seconds: number;
  focus_ratio: number;
}

export const focusApi = {
  start: (taskId?: string) =>
    apiClient<FocusSession>('/focus/sessions', {
      method: 'POST',
      body: JSON.stringify({ task_id: taskId }),
    }),

  getActive: () => apiClient<FocusSession | null>('/focus/active'),

  finish: (sessionId: string, data: { productive_seconds: number; distracted_seconds: number; mark_task_completed?: boolean }) =>
    apiClient<FocusSession>(`/focus/sessions/${sessionId}/finish`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  list: () => apiClient<FocusSession[]>('/focus/sessions'),

  getToday: () => apiClient<FocusSummaryToday>('/focus/today'),
};

File: backend/app/schemas/focus.py
from pydantic import BaseModel, ConfigDict
from uuid import UUID
from datetime import datetime
from typing import Optional
from app.schemas.task import TaskResponse

class FocusSessionStart(BaseModel):
    task_id: Optional[UUID] = None

class FocusSessionUpdate(BaseModel):
    duration_seconds: int
    productive_seconds: int
    distracted_seconds: int
    status: Optional[str] = None

class FocusSessionFinish(BaseModel):
    productive_seconds: int
    distracted_seconds: int
    complete_task: bool = False
    mark_task_completed: bool = False

class FocusSessionResponse(BaseModel):
    id: UUID
    user_id: UUID
    task_id: Optional[UUID] = None
    started_at: datetime
    ended_at: Optional[datetime] = None
    duration_seconds: int
    productive_seconds: int
    distracted_seconds: int
    status: str
    created_at: datetime
    task: Optional[TaskResponse] = None

    model_config = ConfigDict(from_attributes=True)

class FocusSummaryToday(BaseModel):
    total_sessions: int = 0
    total_focus_seconds: int = 0
    focus_seconds: int = 0
    total_distraction_seconds: int = 0
    distracted_seconds: int = 0
    focus_percentage: float = 0.0
    focus_ratio: float = 0.0
    is_active: bool = False
    active_session: Optional[FocusSessionResponse] = None

    model_config = ConfigDict(from_attributes=True)
